README
#
# Copyright (C) 2006-2012 Brian Elliott Finley <brian@thefinleys.com>
#
#
#   Simple State Manager
#
#       SSM (Simple State Manager) is a tool that can be used to ensure
#       consistent state across a large number of machines.  
#
#       SSM includes the following features, plus others:
#
#       - Designed to use a simple, easy to read configuration file
#       - State definition config file supports includes (other config 
#         file chunks)
#       - State definition files may be located on the managed system,
#         or on a central web or ftp server.
#       - Supports priorities (this version of file X has priority over
#         that version)
#       - Handles files and packages (dpkg/apt and rpm/yum based OSes) 
#       - Handles the following file types:
#           - regular
#           - directory
#           - softlink
#           - hardlink
#           - block
#           - character
#           - fifo
#

########################################################################
#
#   This File
#
# This is a heavily commented example state definition file.  It is
# intended to be used as a reference.  It should contain (at least) one
# of each type of configuration entry that can be used in this file.
#
# It can literally be used as a test configuration or to demonstrate SSM
# functionality.
#
########################################################################


########################################################################
#
# [global] settings
#
[global]
#pkg_manager = aptitude
#pkg_manager = dpkg
pkg_manager = none
#pkg_manager = yum
#
#
#   
# NOTE: base_uri defaults to the location where the definition_file
# sits, according to the "definition_file" setting in
# /etc/ssm/client.conf.
#
#base_uri   http://server/OU/FQDN
#base_uri   https://server/OU/FQDN
#base_uri   ftp://server/OU/FQDN
#base_uri    file:///etc/ssm/FQDN
#
#   
#
# upload_url is used when telling ssm to add a file on a system to the
# repository as the new version that should be used.
#
#upload_url = ssh://xcat-master/install/ssm_repo/
#upload_url = ssh://username@xcat-master/install/ssm_repo/
#upload_url = file://install/ssm_repo/
#
#   
#   
# If email_log_to is set, then the log file will be emailed to the
# listed address.
#
#email_log_to    john.doe@example.com
email_log_to    root@localhost
#
#
#
# remove_running_kernel defaults to "no" if not specified.  It can also
# be set to "yes".
#
#remove_running_kernel = no
#
#
#
# Set permissions for resultant log file (/var/log/ssm).  May be useful
# for allowing non-root users to review the logs.  
#
# Defaults to 0640 (-rw-r-----).
#
#log_file_perms  0640

#
########################################################################

########################################################################
#
# [files]
#
#   Each file section allows you to define a desired state for the named
#   file, which includes a variety of critical characterstics that are
#   appropriate based on the type of file.
#
#   File 'type' can be one of:
#       regular
#       directory
#       softlink
#       hardlink
#       block
#       character
#       fifo
#       unwanted
#       ignore
#       generated
#       directory+contents-unwanted
#       chown+chmod
#
#   Optional parameters:
#       
#       depends = /file /directory pkgname1 pkgname2
#
#           * List of packages and/or files which must be installed prior to 
#             taking action on this configuration element.
#           * This is a white space seperated list.
#           * If a list item starts with a slash (/), then it is treated as a
#             file or directory.
#           * If a list item starts with anything other than a slash, it is
#             treated as a package name.
#
#       priority = 0
#           
#           * More than one entry for the same file can be specified.
#             The entry with the higher priority wins.  Allows a machine
#             specific config to override a common config, for example.
#           * Should be a positive or negative integer.
#           * If no priority specified, value defaults to 0.
#
#       prescript = /some/command
#
#           * The prescript is executed verbatim before taking other
#             action on the "name"ed file or directory.
#
#       postscript = /some/command
#
#           * The postscript is executed verbatim after taking other
#             action on the "name"ed file or directory.
#

#
# "regular"
#
#   A regular file.  A normal plain old file.  Not a symlink, directory,
#   fifo, etc.
#
#   Regular files have their md5sum, ownership, and permissions
#   verified.  
#   * If the md5sum doesn't match, the golden file from the $base_uri
#     is installed, and permissions are set.
#   * If the md5sum matches, but the permissions are wrong, then only
#     the permissions are fixed.
#   * Priority not specified, so zero (0) is implied.
#
[file]
name = /tmp/one-regular-file.txt
type = regular
md5sum = 27abe7c7e2423eddec0839a2d0600000
owner = root
group = root
mode = 0644

#
# "regular"
#
#   Another regular file (same file as above, but perhaps in a bundle,
#   with higher priority).
#
#   * Highest priority wins.
#   * No numerical limit.
#   * Use an integer.
#   * If you don't use an integer, the string length of the value you
#     provide will be used as the integer value.  A value of 'one' would
#     result in an integer value of '3' (3 characters in the word 'one').
#
[file]
name = /tmp/one-regular-file.txt
type = regular
md5sum = 27abe7c7e2423eddec0839a2d0633e37
owner = root
group = root
mode = 0644
priority = 95

#
# "regular"
#
#   Yet another regular file with prescript and postscript.
#
#   Note: prescript and postscript are only run if $file needs to be 
#         synced in any way, including permissions or ownership.
#
#[file]
#name = /var/spool/cron/crontabs/root
#type = regular
#md5sum = 91cdcdd73c8e8d0bf26348b4beae6e27
#owner = root
#group = crontab
#mode = 0600
#postscript = /etc/init.d/cron restart
#prescript = echo prescript

#
# "directory"
#
[file]
name = /tmp/uncle-bob
type = directory
owner = root
group = root
mode = 0705

#
# "directory+contents-unwanted"
#
#   Same as a directory, but remove any contents that are not elsewhere
#   in the definition.
#
#   * Create directory and maintain permissions and ownership on the 
#     directory, just like a simple "type = directory".
#   * Plus, any contents of this directory that are not defined 
#     (elsewhere in the configuration) will be marked as unwanted.
#
#   TIP:  You can use the 'ignore' file type to exclude something from
#         the unwanted contents list.
#
[file]
name    = /tmp/monkey_nest
type    = directory+contents-unwanted
owner   = root
group   = root
mode    = 0705


#
# Softlink
#
#   * If a softlink doesn't exist, it is created.
#   * If a softlink points to the wrong target, it is removed and
#     re-created.
#   * If the target doesn't exist, the softlink is still created.
#     A WARNING is indicated, but ERROR_LEVEL is _not_ increased.
#
[file]
name        = /tmp/hosts
type        = softlink
target      = ../etc/hosts

[file]
name        = /tmp/softlink
type        = softlink
target      = /tmp/that/target
prescript   = echo this is the prescript
postscript  = echo this is the postscript

#
# Hardlink
#
#   * If a hardlink doesn't exist, it is created. 
#       * But if the target does not exist, creation of the link is
#         auto-skipped, and $ERROR_LEVEL (which is used as the exit
#         status) is incremented.
#   * If a hardlink exists, the inode of the link and target are
#     compared: 
#       * If they match, nothing is done.
#       * If they don't match, the link is removed, and re-created.
#
[file]
name    = /tmp/ls
type    = hardlink
target  = /bin/busybox

#
# Block, Character, and Fifo
#
#   If a special file exists, and it's major and minor numbers don't
#   match the definition, it is removed and re-created.
#   * If the file type and the major and minor numbers match, but the
#     permissions are wrong, then only the permissions are fixed.
#
[file]
name    = /tmp/null
type    = character
major   = 1
minor   = 3
owner   = root
group   = root
mode    = 0666

[file]
name    = /tmp/sda
type    = block
major   = 8
minor   = 0
owner   = root
group   = disk
mode    = 0660

[file]
name    = /tmp/xconsole
type    = fifo
owner   = root
group   = adm
mode    = 0640

#
# Unwanted file
#
#   Unwanted files or directories simply removed.
#
[file]
name    = /tmp/monkey
type    = unwanted

#
# Unwanted directory
#
[file]
name    = /tmp/monkey_dir
type    = unwanted

#
# Ignore file
#
[file]
name    = /tmp/bob
type    = ignore

#
# "generated"
#
#   A file that is generated via a generator that is any series of
#   commands that can be run from the shell.  The resultant file is
#   literally what you get if you "scrape-n-paste" the commands
#   specified as the "generator".
#
#   * Warning:  generated files have the potential to be an expensive
#     operation, as the "generator" is run every single time the file 
#     is tested.
#
[file]
name        = /tmp/hostname
type        = generated
generator   = hostname -f
owner       = root
group       = root
mode        = 0775
depends     = /bin/hostname

[file]
name        = /tmp/monkey_nest/eggs
type        = generated
generator   = echo "monkey eggs"
owner       = root
group       = root
mode        = 0775

[file]
name = /etc/sysconfig/network-scripts/ifcfg-eth2
type       = generated
generator  = HOSTNAME=$(hostname -s); IP=$(getent hosts $HOSTNAME | awk
'{print $1}'); echo DEVICE=eth2; echo NM_CONTROLLED=no; echo ONBOOT=yes;
echo IPADDR=$IP; echo NETMASK=255.255.255.0; echo BOOTPROTO=static; echo
PEERDNS=no
owner      = root
group      = root
mode       = 0644


#
# "chown+chmod"
#
#   Disregarding file type, simply maintains permissions and ownership.
#
#   * Maintaines ownership and permissions on a file or directory, but
#     disregards contents.
#   * Creates an empty file if file doesn't yet exist.
#
[file]
name = /tmp/monkeyboy
type = chown+chmod
owner = root
group = root
mode = 0642


#
########################################################################


########################################################################
#
# [service] sections
# 
# name:
#   * /etc/init.d/$name is assumed for the init script.
#
# details:
#   * Simply specify the runlevel a colon (:) and the prefix for each
#     service link to the master init script.
#   * SSM takes a look at your client system and gets a list of all of
#     it's rc directories (/etc/rc2.d, /etc/rcS.d, etc.).  
#     * If any rc directory contains a file with a name that matches 
#       (link or not -- some apps install copies instead of links), but
#       is not defined, it will be removed.
#     * Any defined links will be created.
#   * Order of entry of details does not matter.
#
# depends (optional):
#   * List of packages and/or files which must be installed prior to 
#     taking action on this configuration element.
#   * This is a white space seperated list.
#   * If a list item starts with a slash (/), then it is treated as a
#     file or directory.
#   * If a list item starts with anything other than a slash, it is
#     treated as a package name.
#
# Note: a "name that matches" means a filename that starts with either
#       an upper case S or K, followed by two digits, followed by the
#       name of the init script in question.
#
# This example will create the following links, and will remove all
# others from the /etc/rc?.d directories:
#
#   /etc/rc2.d/S99test-service
#   /etc/rc0.d/K01test-service
#   /etc/rcS.d/K01test-service
#   /etc/rc6.d/K01test-service
#
[service]
name = example-service
details = 2:S99 0:K01 6:K01

#[service]
#name = slapd
#details = 2:S99 0:K01 S:K01 6:K01
#depends = slapd /etc/ldap/slapd.conf

#
########################################################################


########################################################################
#
# [bundles]
# 
#   The "bundles" sections (only one is really necessary) allow you to
#   include chunks of definition files, which are typically bundles of
#   related configuration elements.  
#
#   For example, you might have a bundle that includes all of the
#   packages and files necessary for your site's postfix mail client
#   configuration.
#
#   Details:
#
#       * Simply list the filename for each bundle in your bundles
#         section.
#       * Bundle file names may be relative to the "base_uri", or they
#         may be full URIs of type file://, ftp://, http://, or
#         https://.
#       * A bundle can included any non-global configuration elements.
#   
#[bundles]
#bundles/postfix-client
#bundles/kerberos-client
#bundles/web-server

#   
########################################################################


########################################################################
#
# [packages] sections
#
# If no packages are specified, no package related actions will be taken.
#
# Details:

ChangeLog
2007-04-27 19:00  finley

	* VERSION: v0.3.0

2007-04-27 18:59  finley

	* lib/SystemStateManager.pm: Gracefully handle commit of file in
	  new path

2007-04-27 18:28  finley

	* debian/changelog, lib/SystemStateManager.pm,
	  usr/share/doc/examples/one-of-each.conf: Genericize svn_uri;
	  include it in definition file.

2007-04-23 21:29  finley

	* lib/SystemStateManager.pm: Add 'a' option for adding a file to
	  the repo.

2007-04-06 20:16  finley

	* TODO: TODO update

2007-03-23 22:14  finley

	* TODO: Update TODO

2007-03-15 17:14  finley

	* lib/SystemStateManager.pm: * Fix the following bugs when using on
	  non-dpkg systems:
	  * Odd number of elements in hash assignment
	  * Use of uninitialized value in list assignment

2007-03-07 20:41  finley

	* VERSION: up version to 0.2.1

2007-03-07 20:41  finley

	* lib/SystemStateManager.pm: * Handle packages "provided" by
	  packages in state definition
	  * Better output when running pre and post-script commands

2007-02-15 22:07  finley

	* VERSION, debian/changelog: update changelog and VERSION

2007-02-15 22:01  finley

	* TODO, bin/ssm, lib/SystemStateManager.pm: * ADDED: In
	  /etc/ssm/client.conf, if "definition_file" value ends with a
	  slash (/), auto append `hostname -f`
	  * ADDED: Print "Number of errors: N" at end of run.

2007-02-15 20:21  finley

	* TODO, lib/SystemStateManager.pm: * FIXED BUG: Still says it's
	  going to create a directory for a file whos
	  directory doesn't exist.
	  * ADDED: automatically add a file's directory to it's dependency
	  list.

2007-02-15 18:17  finley

	* TODO, VERSION, debian/changelog, lib/SystemStateManager.pm: * Fix
	  spaces between bundle names in display. Again.
	  * If base_uri is not set in the definition file, then do a
	  dirname on the location of the definition file (from the
	  client.conf file) and use that as base_uri.

2007-01-22 20:06  finley

	* TODO: * update TODO

2007-01-10 16:45  finley

	* TODO:

2007-01-10 16:44  finley

	* Makefile: fix shell regex/globs to work properly with 'sh'

2007-01-10 16:42  finley

	* lib/SystemStateManager.pm: Remove 'Please verify owner' bit from
	  ssm_add-file sample entries.

2007-01-05 15:35  finley

	* TODO, VERSION, debian/changelog, lib/SystemStateManager.pm: Use
	  "bu" to backup modified files (if bu exists) prior to installing
	  a file.

2006-12-11 20:35  finley

	* TODO:

2006-12-01 20:12  finley

	* debian/changelog, lib/SystemStateManager.pm: Fix spaces between
	  included bundles.

2006-12-01 19:16  finley

	* VERSION, debian/changelog: VERSION rev.

2006-12-01 19:03  finley

	* TODO, bin/ssm, lib/SystemStateManager.pm: - check first for
	  updates to ssm
	  - install them first, then re-start with new binary

2006-12-01 17:53  finley

	* TODO, lib/SystemStateManager.pm: print out bundles being used.

2006-11-29 04:38  finley

	* TODO, VERSION, debian/changelog, lib/SystemStateManager.pm: Fixed
	  symlink bug.

2006-09-27 16:46  finley

	* VERSION, debian/changelog: * new release

2006-09-27 16:45  finley

	* lib/SystemStateManager.pm,
	  usr/share/doc/examples/one-of-each.conf: * Add directory file
	  type

2006-09-13 17:25  finley

	* VERSION, debian/changelog: v0.1.147

2006-09-13 17:23  finley

	* lib/SystemStateManager.pm: * remove too many blank lines with
	  each get_file()
	  * add a space to the end of bundles (which may not have one to
	  indicate the end of a section)

2006-09-03 05:00  finley

	* VERSION, debian/changelog: * New dev release.

2006-09-03 04:59  finley

	* TODO, lib/SystemStateManager.pm,
	  usr/share/doc/examples/one-of-each.conf: * Added support for
	  'bundles' (include files).

2006-09-01 16:28  finley

	* lib/SystemStateManager.pm: * allow deps for backwards
	  compatibility

2006-09-01 16:26  finley

	* lib/SystemStateManager.pm: * change 'deps' tag to 'depends'

2006-09-01 15:08  finley

	* usr/share/doc/examples/one-of-each.conf: * document 'depends'
	  option.

2006-08-29 19:49  finley

	* VERSION, debian/changelog: * new dev release

2006-08-29 19:48  finley

	* Makefile: * include latest link in Makefile

2006-08-29 19:46  finley

	* TODO, lib/SystemStateManager.pm: * Add check_deps() for files.

2006-08-25 15:22  finley

	* VERSION, debian/changelog:

2006-08-25 15:21  finley

	* lib/SystemStateManager.pm: * Add date stamp to COMMENT in
	  $md5sum.README files
	  * Add "comment = " field to example [file] entries

2006-08-22 16:18  finley

	* VERSION, debian/changelog: * New version

2006-08-22 16:18  finley

	* lib/SystemStateManager.pm: * fix bug: fix ownership and
	  permissions (even though they don't need fixing).

2006-08-17 20:15  finley

	* lib/SystemStateManager.pm: * Add get_rpm_pkgs_currently_installed
	  function.

2006-08-08 22:02  finley

	* VERSION, debian/changelog, lib/SystemStateManager.pm:

2006-08-08 20:02  finley

	* VERSION, debian/changelog:

2006-08-08 19:59  finley

	* VERSION:

2006-08-08 19:59  finley

	* TODO:

2006-08-08 19:56  finley

	* lib/SystemStateManager.pm: * get deps of rdeps and verify that
	  rdeps really are.

2006-08-08 19:40  finley

	* lib/SystemStateManager.pm: * Improve dependency checking.

2006-08-08 16:50  finley

	* lib/SystemStateManager.pm: * comments and such

2006-08-02 06:01  finley

	* lib/SystemStateManager.pm: * add dependency support for services

2006-08-02 05:59  finley

	* TODO:

2006-08-02 05:38  finley

	* lib/SystemStateManager.pm: * add dependency support for services

2006-08-01 21:20  finley

	* Makefile, lib/SystemStateManager.pm,
	  usr/share/doc/examples/one-of-each.conf: * add services
	  capability

2006-08-01 18:39  finley

	* lib/SystemStateManager.pm: * add list of subroutines

2006-08-01 18:38  finley

	* TODO:

2006-08-01 18:37  finley

	* debian/changelog: * arch all

2006-08-01 18:21  finley

	* debian/control: * arch all

2006-07-21 16:52  finley

	* VERSION, debian/changelog: * Update VERSION, debian/changelog,
	  and/or similar files.

2006-07-21 16:52  finley


lib/SystemStateManager/Yum.pm
#  
#   Copyright (C) 2006-2008 Brian Elliott Finley
#
#   $Id: SystemStateManager.pm 234 2008-10-16 02:06:06Z finley $
#    vi: set filetype=perl tw=0:
# 

package SystemStateManager::Yum;

use Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(
                remove_pkgs
                upgrade_pkgs
                install_pkgs
                get_pkgs_that_pkg_manager_says_to_upgrade
                upgrade_ssm
                get_pkgs_provided_by_pkgs_from_state_definition
                get_pkgs_currently_installed
                get_pkg_dependencies
                get_pkg_reverse_dependencies
                get_running_kernel_pkg_name
            );

use strict;
use SystemStateManager qw(ssm_print run_cmd choose_tmp_file);


################################################################################
#
#   This package provides the following functions:
#
#       % egrep '^sub ' lib/SystemStateManager/Yum.pm | perl -pi -e 's/^sub /#   /; s/ {//;' | sort
#
#   get_pkg_dependencies
#   get_pkg_reverse_dependencies
#   get_pkgs_currently_installed
#   get_pkgs_provided_by_pkgs_from_state_definition
#   get_pkgs_that_pkg_manager_says_to_upgrade
#   get_running_kernel_pkg_name
#   install_pkgs
#   remove_pkgs
#   upgrade_pkgs
#   upgrade_ssm
#
#
################################################################################


################################################################################
#
#   Subroutines
#

sub remove_pkgs {
    ssm_print ">>\n>> remove_pkgs()\n" if( $main::o{debug} );

    my @packages = @_;

    my $file = choose_tmp_file();
    open(FILE,">$file") or die("Couldn't open $file for writing");
    foreach my $pkg (@packages) {
        print FILE "remove $pkg\n";
    }
    print FILE "transaction run\n";
    print FILE "exit\n";
    close(FILE);
    my $cmd = "yum -y -C shell $file";
    ssm_print ">> $cmd\n" if( $main::o{debug} );
    ssm_print "FIXING:  Packages -> Removing.\n";
    run_cmd($cmd);

    unlink($file);

    return 1;
}

#XXX is this the right thing to do?  Just an "yum update"?
sub upgrade_pkgs {
    ssm_print ">>\n>> upgrade_pkgs()\n" if( $main::o{debug} );

    my @packages = @_;

    if(scalar(@packages) eq 0) { return 1; }

    ssm_print "FIXING:  Packages -> Upgrading.\n";

    my $file = choose_tmp_file();
    open(FILE,">$file") or die("Couldn't open $file for writing");
    print FILE "update\n";
    print FILE "run\n";
    print FILE "exit\n";
    close(FILE);
    my $cmd = "yum -y shell $file";
    ssm_print ">>   $cmd\n" if( $main::o{debug} );
    run_cmd($cmd);

    unlink($file);

    return 1;    
}


sub install_pkgs {
    ssm_print ">>\n>> install_pkgs()\n" if( $main::o{debug} );

    my @packages = @_;

    if(scalar(@packages) eq 0) { return 1; }

    ssm_print "FIXING:  Packages -> Installing.\n";

    my $file = choose_tmp_file();
    open(FILE,">$file") or die("Couldn't open $file for writing");
    foreach my $pkg (@packages) {
        print FILE "install $pkg\n";
    }
    print FILE "run\n";
    print FILE "exit\n";
    close(FILE);
    my $cmd = "yum -y shell $file";
    ssm_print ">>   $cmd\n" if( $main::o{debug} );
    run_cmd($cmd);

    unlink($file);

    return 1;    
}


sub get_pkgs_that_pkg_manager_says_to_upgrade {
    ssm_print ">>\n>> get_pkgs_that_pkg_manager_says_to_upgrade()\n" if( $main::o{debug} );

    # In this hash, 'pkg' is the key, and 'version' is the value.
    my %hash;
    my $cmd;

    #
    # Get a list of packages that would be upgraded
    $cmd = 'yum check-update';
    ssm_print ">> $cmd\n" if( $main::o{debug} );

    #
    # This is harmless, so we run it even if --dry-run, so that we get
    # the output that is interesting for the rest of the dry run.
    #
    # Example output:                                                          Note - spaces here
    #                                                                          vvvvvvvvvvvvvvvvvv
    #   cups.x86_64                              1:1.2.4-11.18.el5_2.2  updates         
    #   cups-libs.x86_64                         1:1.2.4-11.18.el5_2.2  updates         
    #   dhcpv6-client.x86_64                     1.0.10-4.el5_2.3       updates         
    #   kernel.x86_64                            2.6.18-92.1.13.el5     updates         
    #   kernel-devel.x86_64                      2.6.18-92.1.13.el5     updates         
    #   kernel-headers.x86_64                    2.6.18-92.1.13.el5     updates         
    #   tzdata.noarch                            2008f-3.el5            updates         
    #
    open(OUTPUT,"$cmd|");
        while(<OUTPUT>) {
            if( m/^(\S+)\s+(\S+)\s+updates/ ) {
                #ssm_print ">>   $1\n" if( $main::o{debug} );
                $hash{$1} = 1;
            }
        }
    close(OUTPUT);

    return (keys %hash);
}


sub upgrade_ssm {

    my $pkg = 'ssm.noarch';

    my @pkgs_to_be_upgraded = get_pkgs_that_pkg_manager_says_to_upgrade();
    foreach(@pkgs_to_be_upgraded) {

        if($_ eq $pkg) {

            ssm_print "FIXING:  Upgrading $pkg\n";
            install_pkgs($pkg);

            my $cmd = $main::o{invocation_command};
            ssm_print "Restarting ssm with: $cmd\n";
            exec($cmd) or die("Couldn't exec $cmd");
        }
    }
    return 1;
}

sub get_pkgs_currently_installed {
    ssm_print ">>\n>> get_pkgs_currently_installed()\n" if( $main::o{debug} );

    #
    # returns a hash: package => version
    #

    my %hash;

    if( $main::o{pkg_manager} eq 'yum' ) {

        my $cmd = 'rpm -qa --queryformat "%{NAME}.%{ARCH}\t%{VERSION}\tinstalled\n"';
        ssm_print ">> SYSTEM: $cmd\n" if( $main::o{debug} );
        open(FILE,"$cmd|") or die("couldn't open $cmd for reading");
        while (<FILE>) {
    
                #
                # Only choose packages marked as installed
                #
                # Sample output:
                #
                #   e2fsprogs.x86_64        1.39-8.el5      installed       
                #   e2fsprogs-libs.i386     1.39-8.el5      installed       
                #   e2fsprogs-libs.x86_64   1.39-8.el5      installed       
                #   ed.x86_64               0.2-38.2.2      installed       
                #   eject.x86_64            2.1.5-4.2.el5   installed       
                #   elfutils-libelf.x86_64  0.125-3.el5     installed       
                #   elinks.x86_64           0.11.1-5.1.el5  installed       
                #   ethtool.x86_64          5-1.el5         installed       
                #   expat.i386              1.95.8-8.2.1    installed       
                #   expat.x86_64            1.95.8-8.2.1    installed  
                #
                my ($pkg, $version, $state) = split;
                if( $state eq 'installed' ) {
                    # 
                    # RPM can import GPG keys to validate signatures on
                    # packages.  But why, oh why, Red Hat, do these
                    # imported keys show up as packages with an 
                    # 'rpm --qa' ?!?!?!!!!!  Argh...  
                    #
                    # This skips over them. -BEF-
                    next if($pkg =~ m/^gpg-pubkey\.\(none\)$/);
                    $hash{$pkg} = $version;
                    ssm_print ">>   Currently Installed:  $pkg = $version\n" if( $main::o{debug} );
                }
        }
        close(FILE);
    }

    return %hash;
}


sub get_pkgs_provided_by_pkgs_from_state_definition {
    ssm_print ">>\n>> get_pkgs_provided_by_pkgs_from_state_definition()\n" if( $main::o{debug} );

    my $PKGS_FROM_STATE_DEFINITION = shift;
    return %$PKGS_FROM_STATE_DEFINITION;
}

#
# Returns a hash of packages, with the package names as the keys.
#
# NOTE:  If there are alternates as dependencies, all alternates are returned
#        as the a key in a space seperated string format. -BEF-
#
sub get_pkg_dependencies {
    ssm_print ">>\n>> get_pkg_dependencies()\n" if( $main::o{debug} );

    my @packages = @_;

    my %dependencies;
    my %erasures;

    my $file = choose_tmp_file();
    open(FILE,">$file") or die("Couldn't open $file for writing");
    print FILE "install";
    foreach my $pkg (@packages) {
        print FILE " $pkg";
    }
    print FILE "\n";
    print FILE "transaction solve\n";
    print FILE "exit\n";
    close(FILE);
    my $cmd = "yum shell $file";
    ssm_print ">>   $cmd\n" if( $main::o{debug} );

    open(OUTPUT,"$cmd|") or die;
        while(<OUTPUT>) {
            chomp;
            #
            # Output looks like:
            #   --> Running transaction check
            #   ---> Package nss-tools.x86_64 0:3.12.1.1-1.el5.centos.1 set to be updated
            #   ---> Package kernel.x86_64 0:2.6.18-92.1.1.el5 set to be installed
            #   ---> Package krb5-libs.x86_64 0:1.6.1-25.el5_2.1 set to be updated
            #   ---> Package krb5-workstation.x86_64 0:1.6.1-25.el5_2.1 set to be updated
            #   ---> Package kernel-headers.x86_64 0:2.6.18-92.1.13.el5 set to be updated
            #   ---> Package tzdata.noarch 0:2008f-3.el5 set to be updated
            #   ---> Package pam_krb5.x86_64 0:2.2.14-1.el5_2.1 set to be updated
            #   ---> Package cups-libs.x86_64 1:1.2.4-11.18.el5_2.2 set to be updated
            #   ---> Package kernel-devel.x86_64 0:2.6.18-92.1.1.el5 set to be installed
            #   ---> Package kernel.x86_64 0:2.6.18-92.1.13.el5 set to be installed
            #   ---> Package nss.x86_64 0:3.12.1.1-1.el5.centos.1 set to be updated
            #   ---> Package pam_krb5.i386 0:2.2.14-1.el5_2.1 set to be updated
            #   ---> Package kernel-devel.x86_64 0:2.6.18-92.1.13.el5 set to be installed
            #   ---> Package initscripts.x86_64 0:8.45.19.1.EL-1.el5.centos set to be updated
            #   ---> Package nss.i386 0:3.12.1.1-1.el5.centos.1 set to be updated
            #   ---> Package libtiff.x86_64 0:3.8.2-7.el5_2.2 set to be updated
            #   ---> Package dhcpv6-client.x86_64 0:1.0.10-4.el5_2.3 set to be updated
            #   ---> Package krb5-libs.i386 0:1.6.1-25.el5_2.1 set to be updated
            #   ---> Package cups.x86_64 1:1.2.4-11.18.el5_2.2 set to be updated
            #   --> Finished Dependency Resolution
            #   --> Running transaction check
            #   ---> Package nss-tools.x86_64 0:3.12.1.1-1.el5.centos.1 set to be updated
            #   ---> Package kernel.x86_64 0:2.6.18-92.1.1.el5 set to be installed
            #   ---> Package krb5-libs.x86_64 0:1.6.1-25.el5_2.1 set to be updated
            #   ---> Package krb5-workstation.x86_64 0:1.6.1-25.el5_2.1 set to be updated
            #   ---> Package kernel-headers.x86_64 0:2.6.18-92.1.13.el5 set to be updated
            #   ---> Package tzdata.noarch 0:2008f-3.el5 set to be updated
            #   ---> Package pam_krb5.x86_64 0:2.2.14-1.el5_2.1 set to be updated
            #   ---> Package cups-libs.x86_64 1:1.2.4-11.18.el5_2.2 set to be updated
            #   ---> Package kernel-devel.x86_64 0:2.6.18-92.1.1.el5 set to be installed
            #   ---> Package kernel.x86_64 0:2.6.18-92.1.13.el5 set to be installed
            #   ---> Package nss.x86_64 0:3.12.1.1-1.el5.centos.1 set to be updated
            #   ---> Package pam_krb5.i386 0:2.2.14-1.el5_2.1 set to be updated
            #   ---> Package kernel-devel.x86_64 0:2.6.18-92.1.13.el5 set to be installed
            #   ---> Package initscripts.x86_64 0:8.45.19.1.EL-1.el5.centos set to be updated
            #   ---> Package nss.i386 0:3.12.1.1-1.el5.centos.1 set to be updated
            #   ---> Package libtiff.x86_64 0:3.8.2-7.el5_2.2 set to be updated
            #   ---> Package dhcpv6-client.x86_64 0:1.0.10-4.el5_2.3 set to be updated
            #   ---> Package krb5-libs.i386 0:1.6.1-25.el5_2.1 set to be updated
            #   ---> Package cups.x86_64 1:1.2.4-11.18.el5_2.2 set to be updated
            #   ---> Package kernel-devel.x86_64 0:2.6.18-53.1.21.el5 set to be erased
            #   ---> Package kernel.x86_64 0:2.6.18-53.1.21.el5 set to be erased
            #   --> Finished Dependency Resolution
            #
            # So, we need to produce two sets of results (set to be updated -- such packages
            # should already be in the 'To be upgraded' listing):
            #   dependencies    -> m/set to be installed/ 
            #   erasures        -> m/set to be erased/ 
            if(m/\s+Package\s+(\S+)\s+\S+\s+set to be installed/) {
                $dependencies{$1} = 1;   
            }
            elsif(m/\s+Package\s+(\S+)\s+\S+\s+set to be updated/) {
                # Sometimes a packages is 'set to be updated', but is
                # not yet installed.  Go figure... -BEF-
                $dependencies{$1} = 1;   
            }
            elsif(m/\s+Package\s+(\S+)\s+\S+\s+set to be erased/) {
                $erasures{$1} = 1;   
            }
        }
    close(OUTPUT);
    unlink($file);

    # Dependencies and erasures that are the packages we're getting dependencies for
    # don't count.  Remove 'em.
    #
    # Ok, what's really going on here.  We'll, rather than try to explain it,
    # in the output above, search for /kernel\./ as an example. -BEF-
    foreach my $pkg (@packages) {
        if(defined $dependencies{$pkg}) { delete $dependencies{$pkg}; }
        if(defined $erasures{$pkg})     { delete $erasures{$pkg};     }
    }

    if( $main::o{debug} ) {
        foreach(sort keys %dependencies) {
            ssm_print ">>   Dependency: $_\n";
        }
        foreach(sort keys %erasures) {
            ssm_print ">>   Erasure: $_\n";
        }
    }
    return (\%dependencies, \%erasures);
}


sub get_pkg_reverse_dependencies {
    ssm_print ">>\n>> get_pkg_reverse_dependencies()\n" if( $main::o{debug} );

    my @packages = @_;

    my %reverse_dependencies;

    #
    # Find the reverse dependencies -- packages that depend on the package we're
    # considering removing.
    #
    # Yum doesn't have a way to simply ask it this.  But we can invoke the
    # "yum shell" feature, getting it to calculate the result of our potential 
    # transaction without actually doing it.  That way it will show us the 
    # reverse dependency list we're interested in.
    #
    my $file = choose_tmp_file();
    open(FILE,">$file") or die("Couldn't open $file for writing");
    foreach my $pkg (@packages) {
        print FILE "remove $pkg\n";
    }
    print FILE "transaction solve\n";
    print FILE "exit\n";
    close(FILE);
    my $cmd = "yum -C shell $file";
    ssm_print ">> $cmd\n" if( $main::o{debug} );

    open(OUTPUT,"$cmd|") or die;
    while(<OUTPUT>) {
        chomp;
        #
        # Output looks like:
        #
        #   Setting up Yum Shell
        #   > Setting up Remove Process
        #   Loading mirror speeds from cached hostfile
        #    * base: mirror.anl.gov
        #    * updates: mirror.anl.gov
        #    * addons: mirror.anl.gov
        #    * linus: webapps.inside.anl.gov
        #    * extras: mirror.anl.gov
        #   > --> Running transaction check
        #   ---> Package popt.x86_64 0:1.10.2-48.el5 set to be erased
        #   --> Processing Dependency: libpopt.so.0()(64bit) for package: util-linux
        #   --> Processing Dependency: libpopt.so.0()(64bit) for package: passwd
        #   --> Processing Dependency: libpopt.so.0()(64bit) for package: ntsysv
        #   [snip]
        #   --> Processing Dependency: popt = 1.10.2 for package: rpm
        #   --> Running transaction check
        #   ---> Package ntsysv.x86_64 0:1.3.30.1-2 set to be erased
        #   --> Processing Dependency: ntsysv for package: firstboot-tui
        #   ---> Package rpm-build.x86_64 0:4.4.2-48.el5 set to be erased
        #   ---> Package GConf2.x86_64 0:2.14.0-9.el5 set to be erased
        #   [snip]
        #   etc...
        #
        # What we want to capture is any "Package" that is "set to be erased".
        #
        if(m/Package\s+(\S+)\s+(\S+)\s+set to be erased/) {
            $reverse_dependencies{$1} = 1;
        }
    }
    close(OUTPUT);

    unlink($file);

    foreach my $pkg (@packages) {
        if(defined $reverse_dependencies{$pkg}) { 
            # We don't need to claim one of the listed packages as a reverse
            # dependency of one of the other listed packages. -BEF-
            delete $reverse_dependencies{$pkg};
        } else {
            ssm_print ">>   $pkg rdeps include $1\n" if( $main::o{debug} );
        }
    }

    return %reverse_dependencies;
}

sub get_running_kernel_pkg_name {

    use POSIX;
    my $release = (uname())[2];

    my $running_kernel_pkg_name = `rpm -qf /lib/modules/${release}/kernel`;
    chomp $running_kernel_pkg_name;
    $running_kernel_pkg_name =~ s/:.*//;

    return $running_kernel_pkg_name;
}

#
################################################################################


1;

